# Closed Restaurant Exploration

In another notebook we assigned businesses to one of the ten metro areas and screened out businesses that didn't seem to belong to any of those metros.  This now sits in a nice JSON file bus_metro2.json.

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')
import json
import numpy as np
import pandas as pd

In [2]:
with open("bus_metro2.json", "rt") as infile:
    dfb = pd.read_json(infile)

In [3]:
dfb

attributes  \
--0ZoBTQWQra1FxD4rBWmg                    {'Accepts Credit Cards': False}   
--1emggGHgoG6ipd_RMb-g  {'Accepts Credit Cards': True, 'Price Range': ...   
--4Pe8BZ6gj57VFL5mUE8g  {'Accepts Credit Cards': True, 'Price Range': ...   
--5jkZ3-nUPZxUvtcbr8Uw  {'Waiter Service': False, 'Accepts Credit Card...   
--7PRjnsjMA6uhPK8mW13Q                                                 {}   
--AKjxBmhm9DWrh-e0hTOw  {'Waiter Service': True, 'Price Range': 2, 'De...   
--BlvDO_RG2yElKu9XA1_g  {'Waiter Service': True, 'Accepts Credit Cards...   
--Dl2rW_xO8GuYBomlg9zw                      {'By Appointment Only': True}   
--LzU4UT9rysIDXbhoWHiQ  {'Accepts Credit Cards': True, 'Price Range': ...   
--NCcmQqh_NZbjW1S5XsGQ  {'Price Range': 2, 'Parking': {'valet': False,...   
--Ol5mVSMaW8ExtmWRUmKA  {'Accepts Credit Cards': False, 'By Appointmen...   
--UE_y6auTgq3FXlvUMkbw  {'Accepts Credit Cards': False, 'Good for Kids...   
--XBxRlD92RaV6TyUnP8Ow  {'Accepts Credit Cards': True, 'Price Range': ...   
--Y_2lDOtVDioX5bwF6GIw  {'Coat Check': False, 'Accepts Credit Cards': ...   
--jFTZmywe7StuZ2hEjxyA  {'Accepts Credit Cards': True, 'Good for Kids'...   
--m1g9P1wxNblrLANfVqlA  {'Accepts Credit Cards': True, 'Price Range': ...   
--nQiUBxtpjd_ZBuO_PH6w                    {'Accepts Credit Cards': False}   
--pOlFxITWnhzc7SHSIP0A  {'Coat Check': False, 'Accepts Credit Cards': ...   
--qeSYxyn62mMjWvznNTdg  {'Waiter Service': False, 'Accepts Credit Card...   
--sSW-WY3vyASh_eVPGUAw                     {'Accepts Credit Cards': True}   
--ytaF8wLE4-RsW8Cqx9CQ  {'Accepts Credit Cards': True, 'By Appointment...   
--zgHBiQpr8H2ZqSdGmguQ  {'Waiter Service': True, 'Accepts Credit Cards...   
-024YEtnIsPQCrMSHCKLQw                                                 {}   
-04i_5uaAin8qNeJPD3xZQ                     {'Accepts Credit Cards': True}   
-0Ackw6MF82PXO9f9Jh_Kg                           {'Good for Kids': False}   
-0D_CYhlD2ILkmLR0pBmnA  {'Accepts Credit Cards': True, 'Price Range': ...   
-0GkcDiIgVm0XzDZC8RFOg  {'Coat Check': False, 'Accepts Credit Cards': ...   
-0HGqwlfw3I8nkJyMHxAsQ  {'Waiter Service': False, 'Accepts Credit Card...   
-0OQQRH8MlhpqlUThSAPpg                                                 {}   
-0Oh0BEtQEC9OmmzZ_H5Bg  {'Accepts Credit Cards': True, 'Wi-Fi': 'free'...   
...                                                                   ...   
zyV7n8eXVTAtJrryRswHvA  {'Price Range': 3, 'By Appointment Only': Fals...   
zyZ883vct64bKzNv4wy6Yg  {'Accepts Credit Cards': True, 'Price Range': ...   
zyjyDSdyw0eUqxAWnM_fdA  {'Accepts Insurance': True, 'By Appointment On...   
zypmytGA92yk45AYn60ZHw                     {'Accepts Credit Cards': True}   
zyvkb5nMWUIGo8J2YirW8g  {'Accepts Credit Cards': True, 'By Appointment...   
zyvujnBGGpG3WWiDPkil9w  {'Waiter Service': True, 'Accepts Credit Cards...   
zz4HioU_qObV2kuOsIMUhw  {'Accepts Credit Cards': True, 'Wi-Fi': 'free'...   
zz5eC4ux3D8AOrKKYCU0YA  {'Accepts Credit Cards': True, 'By Appointment...   
zz9dGH66r16YLNaxN80_Bg  {'Waiter Service': False, 'Accepts Credit Card...   
zzGy7cD_5E6sEhs-__u4Pg                    {'Accepts Credit Cards': False}   
zzIP8b0fpMEacKp5KzuLrA  {'Accepts Credit Cards': True, 'Price Range': ...   
zzK7j06Pg8YWkQHzgsTuQA  {'Coat Check': False, 'Accepts Credit Cards': ...   
zzUUj9BUNn0_0F5hwy7RRw  {'Waiter Service': True, 'Accepts Credit Cards...   
zzW_Gu-O5UpieKKp0Vd_bQ  {'Accepts Credit Cards': True, 'Price Range': ...   
zzYURqVx-3W5STDMmh6oxw  {'Accepts Credit Cards': True, 'By Appointment...   
zzZ_JK0TUHSAv-6MxUZj1g  {'Accepts Credit Cards': True, 'By Appointment...   
zze1eI588TqdJLs8xOzMxw  {'Accepts Credit Cards': True, 'Price Range': ...   
zzegBuMzhUosuUcyDNWncQ  {'Accepts Credit Cards': True, 'Delivery': Tru...   
zzgXEteMduykqN8JAUG4CA  {'Accepts Credit Cards': True, 'Caters': True,...   
zzhSCOVTZ7MUKOAfwT-Bsw  {'Accepts Credit Cards': True, 'Price Range': ...   
zzhykRiQh2FyrYEPMfBw0A  {'Accepts Credit Cards': True, 'Pri

Create a dataframe of closed restaurants in Phoenix. Create a second dataframe of open businesses in Phoenix

In [4]:
closed_phx = dfb[(dfb.open == False) & (dfb.categories.map(lambda x: "Restaurants" in x))& (dfb.metro == 'Phoenix')]
print(len(closed_phx))
open_phx = dfb[(dfb.open == True) & (dfb.metro == 'Phoenix')]
print(len(open_phx))

2610
31186


Now we have 2610 closed restaurants and 31186 open businesses. How many open businesses are located at these closed restaurants?

Merge the two dataframes based on the full address with an inner merge, only keeping what shows up in both dataframes

In [5]:
reopen_phx = pd.merge(closed_phx, open_phx, how = 'inner', left_on = 'full_address', right_on = 'full_address')
print(len(reopen_phx))
print(len(reopen_phx.full_address.unique()))

2218
819


In [6]:
reopen_phx.sort_values('name_x')

attributes_x  \
140   {'Waiter Service': False, 'Accepts Credit Card...   
1819  {'Coat Check': False, 'Accepts Credit Cards': ...   
1820  {'Coat Check': False, 'Accepts Credit Cards': ...   
2087  {'Waiter Service': False, 'Accepts Credit Card...   
2086  {'Waiter Service': False, 'Accepts Credit Card...   
1450  {'Waiter Service': True, 'Accepts Credit Cards...   
336   {'Waiter Service': True, 'Accepts Credit Cards...   
1205  {'Waiter Service': True, 'Accepts Credit Cards...   
1043  {'Waiter Service': True, 'Accepts Credit Cards...   
111   {'Waiter Service': False, 'Accepts Credit Card...   
1387  {'Waiter Service': True, 'Accepts Credit Cards...   
1730  {'Coat Check': False, 'Accepts Credit Cards': ...   
2187  {'Waiter Service': True, 'Accepts Credit Cards...   
1315  {'Waiter Service': True, 'Accepts Credit Cards...   
1311  {'Waiter Service': True, 'Accepts Credit Cards...   
1287  {'Waiter Service': True, 'Accepts Credit Cards...   
2123  {'Waiter Service': False, 'Accepts Credit Card...   
1466  {'Waiter Service': True, 'Accepts Credit Cards...   
1465  {'Waiter Service': True, 'Accepts Credit Cards...   
1464  {'Waiter Service': True, 'Accepts Credit Cards...   
1463  {'Waiter Service': True, 'Accepts Credit Cards...   
1462  {'Waiter Service': True, 'Accepts Credit Cards...   
385   {'Waiter Service': True, 'Accepts Credit Cards...   
1556  {'Waiter Service': False, 'Accepts Credit Card...   
1555  {'Waiter Service': False, 'Accepts Credit Card...   
1838  {'Price Range': 1, 'Delivery': True, 'Take-out...   
55    {'Waiter Service': False, 'Accepts Credit Card...   
56    {'Waiter Service': False, 'Accepts Credit Card...   
2113  {'Accepts Credit Cards': True, 'Delivery': Fal...   
1191  {'Accepts Credit Cards': True, 'Good for Kids'...   
...                                                 ...   
538   {'Waiter Service': False, 'Accepts Credit Card...   
539   {'Waiter Service': False, 'Accepts Credit Card...   
540   {'Waiter Service': False, 'Accepts Credit Card...   
541   {'Waiter Service': False, 'Accepts Credit Card...   
542   {'Waiter Service': False, 'Accepts Credit Card...   
543   {'Waiter Service': False, 'Accepts Credit Card...   
544   {'Waiter Service': False, 'Accepts Credit Card...   
545   {'Waiter Service': False, 'Accepts Credit Card...   
547   {'Waiter Service': False, 'Accepts Credit Card...   
546   {'Waiter Service': False, 'Accepts Credit Card...   
1506  {'Waiter Service': False, 'Accepts Credit Card...   
238   {'Waiter Service': False, 'Accepts Credit Card...   
1139  {'Waiter Service': False, 'Accepts Credit Card...   
1138  {'Waiter Service': False, 'Accepts Credit Card...   
1140  {'Waiter Service': False, 'Accepts Credit Card...   
1302  {'Waiter Service': True, 'Accepts Credit Cards...   
2014  {'Waiter Service': True, 'Accepts Credit Cards...   
2015  {'Waiter Service': True, 'Accepts Credit Cards...   
2017  {'Waiter Service': True, 'Accepts Credit Cards...   
2018  {'Waiter Service': True, 'Accepts Credit Cards...   
2019  {'Waiter Service': True, 'Accepts Credit Cards...   
2020  {'Waiter Service': True, 'Accepts Credit Cards...   
2012  {'Waiter Service': True, 'Accepts Credit Cards...   
2011  {'Waiter Service': True, 'Accepts Credit Cards...   
2013  {'Waiter Service': True, 'Accepts Credit Cards...   
2016  {'Waiter Service': True, 'Accepts Credit Cards...   
2138  {'Waiter Service': True, 'Accepts Credit Cards...   
1170  {'Coat Check': False, 'Accepts Credit Cards': ...   
1447  {'Waiter Service': True, 'Accepts Credit Cards...   
987   {'Waiter Service': True, 'Accepts Credit Cards...   

                                           categories_x      city_x  \
140                                [Pizza, Restaurants]        Mesa   
1819  [Bars, American (Traditional), Burgers, Nightl...     Phoenix   
1820  [Bars, American (Traditional), Burgers, Nightl...     Phoenix   
2087                               [Pizza, Restaurants]  Scottsdale   
2086                         

Now there are 2218 businesses that have reopened at addresses of closed restaurants. However, many are duplicate. We see there are 819 businesses at unique addresses, so there should be 2610-819 = 1791 businesses where no business has opened at that address. This also leaves open the possibility that the same building now has a different address.

But really we want the number of unique combinations of address and name.

In [7]:
print(len(reopen_phx.groupby(['full_address','name_x'])))

print(len(closed_phx.groupby(['full_address','name'])))

1014
2606


So there are 2,606 unique combinations of address and business name that are closed in the database. 
There are 1,014 unique combinations of address and business name of closed businesses that match addresses of open businesses.

So, 2606 - 1014 = 1592 is the anticipated number of closed businesses that should still be closed.

How do we get the list of closed restaurants that have stayed closed? Let's try a left merge of the original closed restaurants with the reopened list and filter out anything that doesn't join. 

In [8]:
unopen_phx = pd.merge(closed_phx, reopen_phx, how = 'left', left_on = 'full_address', right_on = 'full_address')
print(len(unopen_phx[unopen_phx.metro_y.isnull()].groupby(['full_address','name'])))


1592


Now we have the anticipated 1,592 records. Let's look at them

In [9]:
unopen_phx[unopen_phx.metro_y.isnull()].sort_values('name')

attributes  \
5592  {'Waiter Service': True, 'Accepts Credit Cards...   
635   {'Waiter Service': True, 'Accepts Credit Cards...   
4205  {'Waiter Service': True, 'Accepts Credit Cards...   
5458  {'Waiter Service': False, 'Accepts Credit Card...   
5304  {'Waiter Service': False, 'Accepts Credit Card...   
3486  {'Waiter Service': False, 'Accepts Credit Card...   
1542  {'Waiter Service': True, 'Accepts Credit Cards...   
3283  {'Waiter Service': True, 'Accepts Credit Cards...   
5572  {'Waiter Service': True, 'Accepts Credit Cards...   
2690  {'Waiter Service': False, 'Accepts Credit Card...   
5528  {'Waiter Service': True, 'Accepts Credit Cards...   
2720  {'Waiter Service': False, 'Order at Counter': ...   
5310  {'Waiter Service': True, 'Accepts Credit Cards...   
236   {'Waiter Service': True, 'Accepts Credit Cards...   
5927  {'Coat Check': False, 'Accepts Credit Cards': ...   
1781  {'Waiter Service': True, 'Accepts Credit Cards...   
288   {'Waiter Service': False, 'Accepts Credit Card...   
2092  {'Accepts Credit Cards': True, 'Price Range': ...   
4926  {'Waiter Service': True, 'Accepts Credit Cards...   
2801  {'Accepts Credit Cards': True, 'Takes Reservat...   
2306  {'Waiter Service': True, 'Accepts Credit Cards...   
928   {'Waiter Service': False, 'Accepts Credit Card...   
2340  {'Waiter Service': True, 'Accepts Credit Cards...   
3450  {'Waiter Service': True, 'Accepts Credit Cards...   
4886  {'Waiter Service': True, 'Accepts Credit Cards...   
1945  {'Waiter Service': True, 'Accepts Credit Cards...   
5442  {'Waiter Service': False, 'Accepts Credit Card...   
1507  {'Waiter Service': False, 'Accepts Credit Card...   
615   {'Coat Check': True, 'Accepts Credit Cards': T...   
3908  {'Waiter Service': False, 'Accepts Credit Card...   
...                                                 ...   
3610  {'Waiter Service': True, 'Accepts Credit Cards...   
2735  {'Waiter Service': True, 'Accepts Credit Cards...   
4947  {'Waiter Service': False, 'Accepts Credit Card...   
486   {'Waiter Service': False, 'Accepts Credit Card...   
5346  {'Waiter Service': True, 'Accepts Credit Cards...   
5868  {'Waiter Service': True, 'Accepts Credit Cards...   
544   {'Accepts Credit Cards': True, 'Outdoor Seatin...   
4368  {'Waiter Service': True, 'Accepts Credit Cards...   
4692  {'Waiter Service': True, 'Price Range': 2, 'De...   
2211  {'Waiter Service': True, 'Accepts Credit Cards...   
5315  {'Waiter Service': True, 'Accepts Credit Cards...   
4043  {'Waiter Service': True, 'Accepts Credit Cards...   
3350  {'Waiter Service': False, 'Accepts Credit Card...   
252   {'Waiter Service': False, 'Accepts Credit Card...   
5390  {'Waiter Service': False, 'Accepts Credit Card...   
4592  {'Waiter Service': False, 'Accepts Credit Card...   
2772  {'Waiter Service': False, 'Accepts Credit Card...   
4273  {'Waiter Service': False, 'Accepts Credit Card...   
4430  {'Waiter Service': True, 'Accepts Credit Cards...   
3826  {'Waiter Service': True, 'Accepts Credit Cards...   
1370  {'Waiter Service': True, 'Accepts Credit Cards...   
5374  {'Waiter Service': False, 'Accepts Credit Card...   
4956  {'Price Range': 2, 'Good for Kids': True, 'Att...   
3525  {'Waiter Service': False, 'Accepts Credit Card...   
220   {'Waiter Service': True, 'Accepts Credit Cards...   
3366  {'Accepts Credit Cards': True, 'Good for Kids'...   
5672                                                 {}   
1746  {'Waiter Service': True, 'Accepts Credit Cards...   
2333  {'Waiter Service': False, 'Accepts Credit Card...   
1371  {'Waiter Service': True, 'Accepts Credit Cards...   

                                             categories             city  \
5592                    [Buffets, Chinese, Restaurants]          Phoenix   
635   [Greek, American (New), Mediterranean, Restaur...          Phoenix   
4205  [Breakfast & Brunch, American (New), Restaurants]          Phoenix   
5458  [Food, Live/Raw Food, Juice Bars & Smoothies, ...         Chandler   
5304  

In [10]:
#unopen_phx[unopen_phx.metro_y.isnull()].plot(x = 'longitude', y = 'latitude', kind = 'scatter', s = 40)
#mplleaflet.display()

Can we compare against the rating?

In [11]:
#unopen_phx[unopen_phx.metro_y.isnull()].plot(x = 'longitude', y = 'latitude', kind = 'scatter', s = 40)
#pizza[pizza.metro == 'Phoenix'].plot(x = 'longitude', y = 'latitude', kind = 'scatter', s = 40, c = 'stars', cmap = 'plasma')
#mplleaflet.show()

Let's create a second target set - poorly-rated, open restaurants

In [12]:
poor = dfb[(dfb.stars <2.5) & (dfb.open == True) & (dfb.categories.map(lambda x: "Restaurants" in x))& (dfb.metro == 'Phoenix')]

In [13]:
#poor.plot(x = 'longitude', y = 'latitude', kind = 'scatter', s = 40)
#mplleaflet.show()